addon/@ember/test-waiters/wait-for.ts

```diff
@@ -96,3 +138,63 @@ function isThenable(
     typeof maybePromise.then === 'function'
   );
 }
+
+function isGenerator(
+  maybeGenerator: any | CoroutineGenerator<unknown>
+): maybeGenerator is CoroutineGenerator<unknown> {
+  // Because we don't have Symbol.iterator in IE11
+  return (
+    typeof maybeGenerator.next === 'function' &&
+    typeof maybeGenerator.return === 'function' &&
+    typeof maybeGenerator.throw === 'function'
+  );
+}
+
+const GENERATOR_WAITER = buildWaiter('@ember/test-waiters:generator-waiter');
+
+function waitForGenerator<T>(
+  generator: CoroutineGenerator<T>,
+  label?: string
+): CoroutineGenerator<T> {
+  GENERATOR_WAITER.beginAsync(generator, label);
+
+  let isWaiting = true;
+  function stopWaiting() {
+    if (isWaiting) {
+      GENERATOR_WAITER.endAsync(generator);
+      isWaiting = false;
+    }
+  }
+
+  return {
+    next(...args) {
+      let hasErrored = true;
+      try {
+        let val = generator.next(...args);
+        hasErrored = false;
+
+        if (val.done) {
+          stopWaiting();
+        }
+        return val;
+      } finally {
+        // If generator.next() throws, we need to stop waiting. But if we catch
+        // and re-throw exceptions, it could move the location from which the
+        // uncaught exception is thrown, interfering with the developer
+        // debugging experience if they have break-on-exceptions enabled. So we
+        // use a boolean flag and a finally block to emulate a catch block.
```

Comment:
For those that are curious, here is a smallish codepen that demos the dev ergonomics issue I was referring to: https://codepen.io/rwjblue/full/PozYKqw.

When you run it, make sure to have DevTools open and "Pause on uncaught exceptions" set in the sources tab. The execution **should** pause in the `someBadFunction` implementation where `throw new Error()` is called, but it is actually pausing inside the `catch` block where we `throw e`.